### PR TITLE
Release v0.0.2

### PR DESCRIPTION
**Pull request by hjaremko**
Changes:
- Handling player input
- Core game mechanics (items, locations, backpack, stats, npcs, custom actions)

### CODE REVIEW

File: src/main/java/pl/uj/io/cuteanimals/model/action/EquipItem.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private List<String> args;
+
+    public EquipItem() {
+        this.args = new ArrayList<>();
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/EquipItem.java:23]: Maybe change to sth like:
```suggestion
            return new Result("You can't equip an item right now");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/GoAction.java [hunk 2]
```diff
@@ -16,30 +19,44 @@
...
+        this.args = args;
     }
 
     @Override
     public IResult execute(ICharacter character) {
-        // TODO: fetch location object matching `where` and move player there,
-        //  if location does not exist return suitable message
-        return new Result("Going to " + where.toString());
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/GoAction.java:43]: Maybe change to sth like:
```suggestion
            return new Result("You can't go somewhere else right now");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/InvestigateAction.java [hunk 3]
```diff
@@ -26,7 +34,15 @@ public void setArgs(List<String> unused) {
 
     @Override
     public IResult execute(ICharacter character) {
-        // TODO: fetch current player location and return its "look around" method
-        return new Result("Looking around null");
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/InvestigateAction.java:38]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for this.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/PickupAction.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    public PickupAction(Map<String, IItem> items) {
+        this.items = items;
+        this.args = new ArrayList<>();
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/PickupAction.java:25]: Maybe change to sth like this:
```suggestion
            return new Result("You can't pick items right now.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/ShowArmor.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import pl.uj.io.cuteanimals.model.Result;
+import pl.uj.io.cuteanimals.model.interfaces.IAction;
+import pl.uj.io.cuteanimals.model.interfaces.ICharacter;
+import pl.uj.io.cuteanimals.model.interfaces.IResult;
+
+public class ShowArmor implements IAction {
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/ShowArmor.java:15]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for this.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/ShowBackpack.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import pl.uj.io.cuteanimals.model.Result;
+import pl.uj.io.cuteanimals.model.interfaces.IAction;
+import pl.uj.io.cuteanimals.model.interfaces.ICharacter;
+import pl.uj.io.cuteanimals.model.interfaces.IResult;
+
+public class ShowBackpack implements IAction {
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/ShowBackpack.java:16]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for this.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/ShowStats.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import pl.uj.io.cuteanimals.model.Result;
+import pl.uj.io.cuteanimals.model.interfaces.IAction;
+import pl.uj.io.cuteanimals.model.interfaces.ICharacter;
+import pl.uj.io.cuteanimals.model.interfaces.IResult;
+
+public class ShowStats implements IAction {
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/ShowStats.java:15]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for this.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (args.size() == 0) {
+            return new Result("You want to talk... who?");
+        }
+
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java:35]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for this.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        var npc =
+                location.getNPCs()
+                        .stream()
+                        .filter(x -> x.getName().equals(args.get(0)))
+                        .collect(Collectors.toList());
+        args.clear();
+
+        return npc.size() >= 1
+                ? new Result(npc.get(0).getQuote())
+                : new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java:47]: Maybe change to sth like:
```suggestion
                : new Result("There is nobody named like that.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/ThrowAwayAction.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    private List<String> args;
+
+    public ThrowAwayAction() {
+        this.args = new ArrayList<>();
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/ThrowAwayAction.java:23]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for that.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/model/action/UnequipItem.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    private List<String> args;
+
+    public UnequipItem() {
+        this.args = new ArrayList<>();
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (!getAcceptableStates().contains(character.getCurrentGameState())) {
+            return new Result("This action cannot be executed now");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/UnequipItem.java:23]: Maybe change to sth like this:
```suggestion
            return new Result("This isn't time for that.");
```
Just to keep immersion.

File: src/main/java/pl/uj/io/cuteanimals/service/GameService.java [hunk 12]
```diff
@@ -2,22 +2,42 @@
...
-        // IResult result = action.execute(CHARACTER_PLACEHOLDER);
-        //
-        // return result.getMessage();
-        return null;
+    @Autowired
+    public GameService(ItemService itemService) {
+        this.itemService = itemService;
+        WorldMap.getInstance().initialize(itemService);
+        // Every new player should spawn with sword
+        player = new Player();
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/service/GameService.java:28]: Add TODO to change this implementation to use repository one day

File: src/main/resources/application-dev.properties [hunk 13]
```diff
@@ -0,0 +1,3 @@
+spring.jpa.hibernate.ddl-auto=create-drop
+debug=true
+
```
Comment (seqre) [inline src/main/resources/application-dev.properties:3]: Remove newline

File: src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            this.armor = item;
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean removeItem(IItem item) {
+        if (weapon == item) {
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java:54]: This is intentional

File: src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            weapon = null;
+            return true;
+        }
+
+        if (armor == item) {
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java:59]: This is intentional

File: src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public String showItems() {
+        return getItems().toString();
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/ArmorBackpack.java:70]: #72

File: src/main/java/pl/uj/io/cuteanimals/model/Backpack.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    @Override
+    public List<IItem> getItems() {
+        return backpack;
+    }
+
+    @Override
+    public boolean putItem(IItem item) {
+        backpack.add(item);
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Backpack.java:20]: What if we want multiple items of the same type? Ex. potions or monster drops

File: src/main/java/pl/uj/io/cuteanimals/model/Backpack.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    @Override
+    public boolean removeItem(IItem item) {
+        backpack.remove(item);
+        return true;
+    }
+
+    @Override
+    public String showItems() {
+        return backpack.toString();
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Backpack.java:31]: #72

File: src/main/java/pl/uj/io/cuteanimals/model/Inn.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public List<NPC> getNPCs() {
+        return null;
+    }
+
+    @Override
+    public List<IEquipment> getItems() {
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Inn.java:42]: I this should be removed in the next version, let's wait for story implemention

File: src/main/java/pl/uj/io/cuteanimals/model/NPC.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+        return null;
+    }
+
+    @Override
+    public void changeLocation(ILocation where) {}
+
+    @Override
+    public GameState getCurrentGameState() {
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/NPC.java:52]: Next version, please add issue

File: src/main/java/pl/uj/io/cuteanimals/model/Player.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Override
+    public IAttributes getAttributes() {
+        return stats;
+    }
+
+    @Override
+    public Result use(IAction action) {
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Player.java:30]: Use is not desinged yet, to do in v0.1.0

[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Player.java:30]: #92

File: src/main/java/pl/uj/io/cuteanimals/model/NPC.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/NPC.java:52]: #95

File: src/main/java/pl/uj/io/cuteanimals/model/PlayerAttributes.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        this.attack += attack;
+    }
+
+    public void addLevel(int level) {
+        this.level += level;
+    }
+
+    public void addDefence(int defence) {
+        this.defence += defence;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/PlayerAttributes.java:76]: Just use nagative numbers

[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/PlayerAttributes.java:76]: Some Items can have negative stats, this code is more general

File: src/main/java/pl/uj/io/cuteanimals/model/Town.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    public List<NPC> getNPCs() {
+        return npcList;
+    }
+
+    @Override
+    public List<IEquipment> getItems() {
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (hjaremko) [reply src/main/java/pl/uj/io/cuteanimals/model/Town.java:45]: #94

File: src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java (hunk 9 above)
Comment (seqre) [reply src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java:47]: You've changed wrong line xD

File: src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+    public TalkAction(List<String> args, ILocation location) {
+        this.args = args;
+        this.location = location;
+    }
+
+    @Override
+    public IResult execute(ICharacter character) {
+        if (args.size() == 0) {
+            return new Result("There is nobody named like that.");
```
Comment (seqre) [inline src/main/java/pl/uj/io/cuteanimals/model/action/TalkAction.java:31]:
```suggestion
            return new Result("Who do you want to talk to?");
```

**Pull request merged**